postgres_ffi/src/xlog_utils.rs

```diff
@@ -292,3 +294,49 @@ impl XLogRecord {
         self.xl_info == pg_constants::XLOG_SWITCH && self.xl_rmid == pg_constants::RM_XLOG_ID
     }
 }
+
+#[allow(non_upper_case_globals)]
+pub const SizeOfXLogRecord: usize = 24;
+
+// From PostgreSQL headers
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct XLogPageHeaderData {
+    pub xlp_magic: u16,      /* magic value for correctness checks */
+    pub xlp_info: u16,       /* flag bits, see below */
+    pub xlp_tli: TimeLineID, /* TimeLineID of first record on page */
+    pub xlp_pageaddr: u64,   /* XLOG address of this page */
+    pub xlp_rem_len: u32,    /* total len of remaining data for record */
+}
+
+// FIXME: this assumes MAXIMUM_ALIGNOF 8. There are 4 padding bytes at end
+#[allow(non_upper_case_globals)]
+pub const SizeOfXLogShortPHD: usize = 2 + 2 + 4 + 8 + 4 + 4;
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct XLogLongPageHeaderData {
+    pub std: XLogPageHeaderData, /* standard header fields */
+    pub xlp_sysid: u64,          /* system identifier from pg_control */
+    pub xlp_seg_size: u32,       /* just as a cross-check */
+    pub xlp_xlog_blcksz: u32,    /* just as a cross-check */
+}
+
+// FIXME: this assumes MAXIMUM_ALIGNOF 8.
+#[allow(non_upper_case_globals)]
+pub const SizeOfXLogLongPHD: usize = (2 + 2 + 4 + 8 + 4) + 4 + 8 + 4 + 4;
+
+pub fn encode_xlog_long_phd(hdr: XLogLongPageHeaderData) -> Bytes {
+    let b: [u8; XLOG_SIZE_OF_XLOG_LONG_PHD];
+    b = unsafe {
+        std::mem::transmute::<XLogLongPageHeaderData, [u8; XLOG_SIZE_OF_XLOG_LONG_PHD]>(hdr)
+    };
+    Bytes::copy_from_slice(&b[..])
+}
```

Comment:
Should this be using `serde`? Why are we are adopting the native byte order + memory layout as a serialization format?

Also, any use of `unsafe` should have a big `SAFETY` comment explaining why it is necessary, and correct.

Reply:
`XLogPageHeaderData` currently requires 8-byte alignment but ends on a 4-byte boundary, so it adds 4 bytes of padding when included in `XLogLongPageHeaderData`.

What was intended here? We should add comments when it's not obvious.

This code seems fragile (and maybe broken on other architectures). Maybe we should use `serde` if we intend to remove the padding, or make the padding field explicit, so that it's obvious what is intended.

Reply:
This Rust structure should map on correspondent C stricture: provide the same alignment and byte order.
Obviously #[repr(C)]  +  std::mem::transmute can guarantee it much better than serde. Also I have not measured overhead of serde but looking at deep stacks (i debug mode) makes me think that it will be essential.

Concerning alignment of XLogPageHeaderData: this structure were defined in Postgres in this way. And there is  comment explaining calculation of size of this struct.

But I think this code should be rewritten now using FFI: there is no need to define this struct ourselves, lets FFI do this work for us.

Reply:
I'd like to point out that, strictly speaking, reads of uninitialized memory (e.g. padding) are UB in Rust, at least according to [miri](https://github.com/rust-lang/miri), which is the only (automated) way to prevent such violations in the absence of the language standard.

Which is to say that this snippet

```rust
type TimeLineID = u32;

#[repr(C)]
#[derive(Debug, Default)]
struct XLogPageHeaderData {
    pub xlp_magic: u16,
    pub xlp_info: u16,
    pub xlp_tli: TimeLineID,
    pub xlp_pageaddr: u64,
    pub xlp_rem_len: u32,
}

#[repr(C)]
#[derive(Debug, Default)]
struct XLogLongPageHeaderData {
    pub std: XLogPageHeaderData,
    pub xlp_sysid: u64,
    pub xlp_seg_size: u32,
    pub xlp_xlog_blcksz: u32,
}

fn main() {
    let hdr = XLogLongPageHeaderData::default();
    let bytes: [u8; std::mem::size_of::<XLogLongPageHeaderData>()] =
        unsafe { std::mem::transmute(hdr) };
    println!("{:?}", bytes);
}
```

crashes with the following message

```
$ cargo miri run
error: Undefined Behavior: using uninitialized data, but this operation requires initialized memory
   --> /home/dmitry/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/fmt/num.rs:462:5
    |
462 | /     impl_Display!(
463 | |         i8, u8, i16, u16, i32, u32, i64, u64, usize, isize
464 | |             as u64 via to_u64 named fmt_u64
465 | |     );
    | |______^ using uninitialized data, but this operation requires initialized memory
    |
    = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
    = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
...
```

Luckily, we can mitigate this problem by adding an explicit padding field, i.e.

```rust
#[repr(C)]
#[derive(Debug)]
struct XLogLongPageHeaderData {
    pub std: XLogPageHeaderData,

    pub _padding_0: u32, // <= Here

    pub xlp_sysid: u64,
    pub xlp_seg_size: u32,
    pub xlp_xlog_blcksz: u32,
}
```

My gut feeling tells me that UB in Rust is much scarier than its counterpart from C and C++, since Rust has stricter aliasing rules etc.

Reply:
> My gut feeling tells me that UB in Rust is much scarier than its counterpart from C and C++

I think this is correct: it's much easier to have catastrophic bugs due to innocent-looking unsafe blocks.

Good idea running Miri against this! I certainly think our code should be able to run cleanly on Miri, so there's yet another reason not to use `transmute` this way.

I tried out `zerocopy` for just a few minutes, and it appears to do exactly what is wanted here. Except it will refuse to compile on structs with padding; I needed to make the padding explicit to make it work.

However, the problems go beyond just `serde` vs `zerocopy` vs unsafe `transmute`: we really shouldn't have to manually compute the size of various structs and manually compute offsets, anywhere outside of the `postgres_ffi` crate. We really need an abstraction that can hide all of the low-level details.

Reply:
> This Rust structure should map on correspondent C stricture: provide the same alignment and byte order.

So you are saying that postgres compiled on different platforms will write incompatible files? Native byte order and padding rarely make a good file format (or network protocol).

> Obviously #[repr(C)] + std::mem::transmute can guarantee it much better than serde.

`transmute` can lead to undefined behavior with these structs. That's not permissible.

I'm interested in arguing about `serde` any more. It's the idiomatic, correct way to get portable serialization in Rust. The only reason I would even consider not using it here is because people are hinting that this data structure is deliberately non-portable.

> Concerning alignment of XLogPageHeaderData: this structure were defined in Postgres in this way. And there is comment explaining calculation of size of this struct.

The comment is vague. I would expect it to explain whether this struct is supposed to be non-portable due to padding and endian issues.

> But I think this code should be rewritten now using FFI: there is no need to define this struct ourselves, lets FFI do this work for us.

Unfortunately, `bindgen` has a weakness here: it does not allow us to add arbitrary `#[derive(X, Y)]` attributes to auto-generated data structures. See the discussion on #6.
At this point, our only options seem to be:
a) Try to add the "add custom derive" feature to `bindgen` ourselves
b) Write a post-processing script to add the necessary `derive` attributes
c) Use `bindgen` to provide an initial set of structs, then copy/paste them and add the things we need (derives, explicit padding, etc). Maybe add unit tests to verify that our copy-pasted versions remain compatible with what `bindgen` can compute from the latest source code.

Reply:
> So you are saying that postgres compiled on different platforms will write incompatible files?

Yeah, that's how it works at the moment. Surprisingly, [the relevant piece of doc](https://www.postgresql.org/docs/current/storage.html) doesn't seem to state that (please, prove me wrong).

Reply:
@ericseppanen Nonetheless, you might find [this article](https://postgrespro.com/blog/pgsql/5967858) interesting.

Reply:
I opened #207 to figure out how to resolve the `transmute` issue, since this PR is just copying a pattern that already exists.

Reply:
> So you are saying that postgres compiled on different platforms will write incompatible files?
Yes, certainly. Postgres never claim binary compatibility of data files between different platforms. You can not take database from Sparc server, copy it to Intel server and use there.

> Native byte order and padding rarely make a good file format (or network protocol).
Yes, it is not good for protocol (and this is why libpq is using network byte format) but most of DBMSes are writing data to the disk in native format.

> transmute can lead to undefined behavior with these structs. That's not permissible.
As far as I undertstand the only concern is garbage in alignment holes. It usually solved by zeroing all object space. I wonder if it is not done in Rust... 

> The comment is vague. I would expect it to explain whether this struct is supposed to be non-portable due to padding and endian issues.
Yes, this structure is non-portable due to padding and endian issues.